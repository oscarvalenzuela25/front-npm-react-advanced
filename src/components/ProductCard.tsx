import styles from '../styles/styles.module.css';
import useProduct from '../hooks/useProduct';
import React, { createContext, CSSProperties } from 'react';
import {
  ProductContextProps,
  Product,
  OnChangesArgs,
  InitialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces';

export interface Props {
  // Patrones de componentes
  // children: ReactElement | ReactElement[];
  product: Product;
  className?: string;
  style?: CSSProperties;
  onChange?: ({ product, count }: OnChangesArgs) => void;
  value?: number;

  //State initializer
  children: (args: ProductCardHandlers) => JSX.Element;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
