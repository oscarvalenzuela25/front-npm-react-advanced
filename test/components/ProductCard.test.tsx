import React from 'react';
import rendered from 'react-test-renderer';
import ProductCard from '../../src/components/ProductCard';
import { product2 } from './data/products';

describe('ProductImage', () => {
  test('Debe mostrar el componente con una img', () => {
    const wrapper = rendered.create(
      <ProductCard product={product2}>{() => <h1>Titulo</h1>}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
