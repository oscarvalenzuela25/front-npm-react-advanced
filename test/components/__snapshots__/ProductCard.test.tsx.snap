// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente con una img 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Titulo
  </h1>
</div>
`;
